File: python/ch13/fenwick.py
# python/ch13/fenwick.py
"""Fenwick Tree と区間和の全走査ベンチマーク."""
from __future__ import annotations

import time
from typing import List


class Fenwick:
    def __init__(self, n: int) -> None:
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, i: int, delta: int) -> None:
        i += 1
        while i <= self.n:
            self.bit[i] += delta
            i += i & -i

    def prefix(self, i: int) -> int:
        i += 1
        s = 0
        while i > 0:
            s += self.bit[i]
            i -= i & -i
        return s

    def range_sum(self, l: int, r: int) -> int:
        if l == 0:
            return self.prefix(r)
        return self.prefix(r) - self.prefix(l - 1)


def naive_range_sum(arr: List[int], l: int, r: int) -> int:
    return sum(arr[l : r + 1])


def benchmark(n: int = 8000, queries: int = 40000) -> None:
    arr = [1] * n
    ops = [(i % n, (i * 7) % n, (i * 13) % n) for i in range(queries)]

    t0 = time.perf_counter()
    total_naive = 0
    for l, r, delta in ops:
        arr[l] += delta % 3
        lo, hi = min(l, r), max(l, r)
        total_naive += naive_range_sum(arr, lo, hi)
    naive_ms = (time.perf_counter() - t0) * 1000

    fw = Fenwick(n)
    for i, v in enumerate(arr):
        fw.add(i, v)
    t0 = time.perf_counter()
    total_fw = 0
    for l, r, delta in ops:
        fw.add(l, delta % 3)
        lo, hi = min(l, r), max(l, r)
        total_fw += fw.range_sum(lo, hi)
    fw_ms = (time.perf_counter() - t0) * 1000

    print(f"n={n} queries={queries}")
    print(f"naive_range_sum_ms={naive_ms:.2f} checksum={total_naive}")
    print(f"fenwick_ms={fw_ms:.2f} checksum={total_fw}")
    print(f"speedup={naive_ms / max(fw_ms, 1e-9):.1f}x")


def main() -> None:
    benchmark()


if __name__ == "__main__":
    main()

File: go/ch13/fenwick.go
// go/ch13/fenwick.go
package main

import (
	"fmt"
	"time"
)

type Fenwick struct {
	n   int
	bit []int
}

func NewFenwick(n int) *Fenwick {
	return &Fenwick{n: n, bit: make([]int, n+1)}
}

func (f *Fenwick) Add(i, delta int) {
	i++
	for i <= f.n {
		f.bit[i] += delta
		i += i & -i
	}
}

func (f *Fenwick) Prefix(i int) int {
	i++
	s := 0
	for i > 0 {
		s += f.bit[i]
		i -= i & -i
	}
	return s
}

func (f *Fenwick) RangeSum(l, r int) int {
	if l == 0 {
		return f.Prefix(r)
	}
	return f.Prefix(r) - f.Prefix(l-1)
}

func naiveRangeSum(arr []int, l, r int) int {
	s := 0
	for i := l; i <= r; i++ {
		s += arr[i]
	}
	return s
}

func benchmark() {
	n, queries := 8000, 40000
	arr := make([]int, n)
	for i := range arr {
		arr[i] = 1
	}
	ops := make([][3]int, queries)
	for i := 0; i < queries; i++ {
		ops[i] = [3]int{i % n, (i * 7) % n, (i * 13) % n}
	}

	start := time.Now()
	totalNaive := 0
	for _, op := range ops {
		arr[op[0]] += op[2] % 3
		lo, hi := op[0], op[1]
		if lo > hi {
			lo, hi = hi, lo
		}
		totalNaive += naiveRangeSum(arr, lo, hi)
	}
	naiveMs := time.Since(start).Seconds() * 1000

	fw := NewFenwick(n)
	for i, v := range arr {
		fw.Add(i, v)
	}
	start = time.Now()
	totalFw := 0
	for _, op := range ops {
		fw.Add(op[0], op[2]%3)
		lo, hi := op[0], op[1]
		if lo > hi {
			lo, hi = hi, lo
		}
		totalFw += fw.RangeSum(lo, hi)
	}
	fwMs := time.Since(start).Seconds() * 1000

	fmt.Printf("n=%d queries=%d\n", n, queries)
	fmt.Printf("naive_range_sum_ms=%.2f checksum=%d\n", naiveMs, totalNaive)
	fmt.Printf("fenwick_ms=%.2f checksum=%d\n", fwMs, totalFw)
	fmt.Printf("speedup=%.1fx\n", naiveMs/max(fwMs, 1e-9))
}

func main() {
	benchmark()
}
